File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.cpp
// FORCE COMPILER OPTIMIZATIONS
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx2,bmi,bmi2,lzcnt,popcnt")

#include <vector>

using namespace std;

// GLOBAL STATIC BUFFERS (Instant Access, No Allocation Overhead)
// Size 100,005 covers the N=10^5 constraint safely.
// We reuse these for all test cases.
static long long q_val[100005];
static int q_idx[100005];

// IO SPEED HACK (Disables C++ stream syncing)
static const auto _ = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return nullptr;
}();

class Solution {
public:
    int shortestSubarray(vector<int>& nums, int k) {
        // 1. Localize variables for Register caching
        int n = nums.size();
        
        // 2. Direct Pointer Access (Bypasses Vector Operator[] overhead)
        const int* raw_nums = nums.data();

        int min_len = n + 1;
        long long current_sum = 0;

        // 3. Manual Deque via Pointers/Indices
        // head: front of deque, tail: end of deque
        int head = 0;
        int tail = 0;

        // Initialize Deque with (sum=0, index=-1)
        // This handles subarrays starting from index 0.
        q_val[0] = 0;
        q_idx[0] = -1;
        tail = 1;

        // 4. Main Loop
        for (int i = 0; i < n; ++i) {
            // Update prefix sum
            current_sum += raw_nums[i];

            // OPTIMIZATION: Calculate target once per iteration
            // We are looking for: current_sum - old_sum >= k
            // Which means: old_sum <= current_sum - k
            long long target = current_sum - k;

            // PHASE 1: CHECK VALIDITY (Shrink from Head)
            // Condition: q_val[head] <= target
            // Using a pre-calculated 'target' saves 1 subtraction per loop
            while (head < tail && q_val[head] <= target) {
                int len = i - q_idx[head];
                if (len < min_len) min_len = len;
                head++;
            }

            // PHASE 2: MAINTAIN MONOTONICITY (Shrink from Tail)
            // We want q_val to be increasing.
            // If current_sum <= q_val[tail-1], the old value is useless.
            while (head < tail && current_sum <= q_val[tail - 1]) {
                tail--;
            }

            // PHASE 3: PUSH BACK
            q_val[tail] = current_sum;
            q_idx[tail] = i;
            tail++;
        }

        return min_len <= n ? min_len : -1;
    }
};

File: 0862-shortest-subarray-with-sum-at-least-k/0862-shortest-subarray-with-sum-at-least-k.rs
impl Solution {
    pub fn shortest_subarray(nums: Vec<i32>, k: i32) -> i32 {
        let n = nums.len();
        let k = k as i64;
        let mut min_len = (n + 1) as i32;

        let mut q_vals = Vec::<i64>::with_capacity(n + 1);
        let mut q_idxs = Vec::<i32>::with_capacity(n + 1);

        unsafe {
            let q_vals_ptr = q_vals.as_mut_ptr();
            let q_idxs_ptr = q_idxs.as_mut_ptr();
            
            let mut head = 0;
            let mut tail = 0;

            *q_vals_ptr = 0;
            *q_idxs_ptr = 0; 
            tail += 1;

            let mut curr_sum: i64 = 0;

            for i in 0..n {

                curr_sum += *nums.get_unchecked(i) as i64;
                let curr_idx = (i + 1) as i32;

                while head < tail {
                    let val = *q_vals_ptr.add(head);
                    if curr_sum - val >= k {
                        let idx = *q_idxs_ptr.add(head);
                        let len = curr_idx - idx;
                        if len < min_len {
                            min_len = len;
                        }
                        head += 1;
                    } else {
                        break;
                    }
                }

                while head < tail {
                    let val = *q_vals_ptr.add(tail - 1);
                    if curr_sum <= val {
                        tail -= 1;
                    } else {
                        break;
                    }
                }

                *q_vals_ptr.add(tail) = curr_sum;
                *q_idxs_ptr.add(tail) = curr_idx;
                tail += 1;
            }
        }

        if min_len > n as i32 { -1 } else { min_len }
    }
}